doc/specs/#674 - PowerShell cmdlets.md

```diff
@@ -0,0 +1,434 @@
+are considering using this framework to help model the syntax and create a module so users can experiment with the syntax.
+Several of the winget.exe commands provide output that is not representative of properly formatted objects, so there will
+likely be gaps to address natively in the COM interface for the Windows Package Manager, or we will provide an argument to output the results from the client to JSON.
+
+## Inspiration
+
+@ThomasNieto was the first to create an Issue for [Native PowerShell Support](https://github.com/microsoft/winget-cli/issues/221).
+This is one of the [most desired features](https://github.com/microsoft/winget-cli/issues?page=1&q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc+label%3AIssue-Feature) for the Windows Package Manager.
+
+## Solution Design
+
+The initial solution will start with [PowerShell Crescendo Preview.3](https://devblogs.microsoft.com/powershell/announcing-powershell-crescendo-preview-3/).
+Many of the major considerations are related to proper PowerShell nouns and verbs. @MilkFather suggested [command aliases](https://github.com/microsoft/winget-cli/issues/380) for the winget.exe
+client and @Jaifroid also suggested [find](https://github.com/microsoft/winget-cli/issues/1299) rather than search.
+
+## UI/UX Design
+
+The main UI/UX differences will be in the command names, parameters, and how the output is formatted for users. For the sake of
+consistency the sub headings below will map to existing commands in winget.exe to begin drafting some syntax proposals. In order
+to keep the discussion focussed, I'll create GitHub discussions for each of these and link them here as we work through driving
+consensus.
+
+The PowerShell module name will be "WindowsPackageManager"
+The Adjective "WinGet" may change, but is used consistently in the cmdlets until such time as we agree on the proper term.
+
+>Note: Naming things is hard, and one of the conventions of PowerShell is to make nouns singular.
+---
+
+### Client command `winget install`
+
+**PowerShell Cmdlet**
+
+```PowerShell
+Install-WinGetPackage
+```
+**Parameters**
+
+-Manifest
+
+-PackageIdentifier
+
+-Name
+
+-Moniker
+
+-Version
+
+-Source
+
+-Scope
+
+-Exact
+
+-Interactive
+
+-Silent
+>Note: we may switch the default to "silent" from "silent with progress" in winget.exe so there may be changes here.
+
+-Locale
+
+-Log
+
+-Override
+
+-Location
+
+---
+
+### Client command `winget show`
+
+**PowerShell Cmdlet**
+```PowerShell
+Get-WinGetPackageVersion
+```
+
+-Detail
+
+-All
+
+-Version
+
+**Examples**
+
+>Note: The current behavior for `winget show vscode` would look like the example below.
+
+```PowerShell
+Find-WinGetPackage -Moniker vscode | Get-WinGetPackageVersion -Detail
+```
+
+>Note: The current behavior for `winget show vscode --versions` would look like the example below.
+```PowerShell
+Find-WinGetPackage -Moniker vscode | Get-WinGetPackageVersion -All
+```
```

Comment:
This is valid PowerShell code but needlessly complicated from a user experience perspective. Many users will expect these actions to be done with a single command. It'll be interesting to see what feedback you'll get here :-)

Reply:
Find-Package should, uhh, find packages and return all the information. C/F Find-Module. Please do not introduce needless complexity where it can be avoided.

Reply:
I disagree. AFAIK winget repositories use a cached index to speed up searches. It makes a ton of sense to return only the minimal, cached set of data first and then get all the information (which is an expensive operation) only for the package you're really interested in. Other existing cmdlets such as Get-ADUser and Get-WindowsDriver take a similar approach for performance and server-load reasons.

Reply:
If the approach is to return a minimal amount of properties, then two things:
1. The Find-WinGetPackage should return enough properties, by default, to be useful. As you mantion, like Get-ADuser etc.
2. If there are more properties not returned by default - then use a -Property parameter to specify which optional properties to return. 
In this case, I'd argue that the version of a package is important and probably should be returned by default. 

There are two issues therefore: what properties get returned by default and how to get more. I do not support adding cmdlets in the proposed manner.

> Just checking, since it's not in this spec file, these cmdlets will be targeting Windows PowerShell (5.1), not PowerShell Core, right? We can't guarantee that PowerShell Core is installed since it's not shipped with Windows (and it isn't usually installed on systems that aren't used for development unless you work at a really hip company).

I sure hope not. It would be crazy, this long after PowerShell 7 has shipped to be so backwards as to ONLY target WIndows PowerShell.

Reply:
> I sure hope not. It would be crazy, this long after PowerShell 7 has shipped to be so backwards as to ONLY target WIndows PowerShell.

I'd disagree. For all that Microsoft want PowerShell 7 to be the future, I don't know a single enterprise that's using it, precisely because it's not a backwards compatible, drop in replacement for PowerShell 5. In every current, supported version of Windows, PowerShell 7 is a side by side addon and the chance of MS actually ever releasing an in place upgrade from PS5 to PS7+ is slim to none, so PowerShell 5 is going to continue to be what is used by enterprise admins for years to come. 
So yeah, IMHO this module absolutely needs to be PS 5.1 compatible.

Reply:
I would agree that the module should work in Windows PowerShell - I am sorry if you thought I meant otherwise. Probably targeting Windows PowerShell 5.1.

But it must also work, and be validated, using PowerShell 7. I expect the Winget team to do NOTHING to stop it from being fully natively in PowerShell 7 as well. We don't need more teams at Microsoft ignoring PowerShell 7 if it can be avoided.

As for your comments - there is a great deal of backwards compatibility built-in. The Compatibility mechanism works well for the Modules shipped, in particular, by the Windows Server team (for the most part). Some teams have taken the opportunity to update their modules to be .NET Core compatible (c/f ActiveDirectory module). I am just wanting to make sure that the winget team does not deliver a PowerShell module that _only_ runs in Windows PowerShell (c/f WSUS, BestPractices, and the PSScheduledJob modules) and has to be blocked in PowerShell 7.  If the Winget team use APIs that are supported both in .NET and the .NET Framework then I can have it in PowerShell 7, and you can use it with  Windows PowerShell without issue. Seems to me that in 2021, this should be a design goal for this and all new modules.

Reply:
Fair play, I had misread your post as wanting it to not be PS5 compatible. I've got absolutely no objection to things being both PS5 and PS7 compatible, but this tool could be of great use to those who work in enterprise environments where hell is likely to freeze over before PS 7 is widely deployed.

Reply:
I've been engaging the PowerShell team in the discussions with our design, and the crescendo team to aid in rapid prototyping. We will follow their guidance with respect to which versions of PowerShell behave differently so we can maintain the greatest compatibility level. I am certainly not a PowerShell expert, so I'm learning from the experts on that front. I wanted to get this draft out quickly so we could get community feedback. I'm building on the list of concerns brought up here so they can be clearly addressed in the document.

Reply:
@justsomeguynz I agree with your point that there are companies that will never allow PS 7 to be deployed. I do not get the reason to be honest. I consult customers to deploy it widely just for the same of improved commandlets like test-connection etc. 
It's an MSI package, easy to deploy with a small mst. What is their issues with PS 7? Do you have more information about it? I just know it is a controversy because of patchmanagement.org and also some ConfigMgr MEMCM pals.

Reply:
It's not a drop in replacement, and it breaks backwards compatibility. In my experience enterprise IT is pretty resistant to  change that breaks backwards compatibility and unforced changes lol.

Reply:
> I've been engaging the PowerShell team in the discussions with our design, and the crescendo team to aid in rapid prototyping. We will follow their guidance with respect to which versions of PowerShell behave differently so we can maintain the greatest compatibility level. I am certainly not a PowerShell expert, so I'm learning from the experts on that front. I wanted to get this draft out quickly so we could get community feedback. I'm building on the list of concerns brought up here so they can be clearly addressed in the document.

This is excellent news and much to be welcomed. Knowing what I do, Crescendo should provide a good way forward and helps to move this forward and makes the best out of the decision to do this all in C++. But we are where we are, and Crescendo does look promising as a way forward. It also gives the winget team to re-write certain cmdlets in C# where that makes business sense. 

As I stated a while ago, creating the object model is very important. With Crescendo, as I understand it, you use regular expressions to create objects that the Crescendo-generated cmdlet emits. As I have said elsewhere, this is going to be a very difficult task to do well and completely - string to object translation is not an exact science. Thus my first concern is the objects emitted. It should be a goal for winget-ps to emit objects that are as rich and functional as those emitted by other modules and usable in rich cross-silo scenarios. 

The second concern is over verb usage. As mentioned a long time ago, the subcommand of winget do not line up completely with PowerShell standard verb naming.  For example, the show command in winget should probably be get. And hash should be something else. I would like to see the command to verb translation table - and please, please, please only uses approved verbs in all cmdlets. 

I look forward to seeing how this progresses.

Reply:
I've been going through the process of learning the PowerShell verbs and continuing to expand the draft to others so we can get a sense of the best verbs and behaviors. First impressions are very valuable in this area. There is definitely some impedance to overcome between the .exe cli syntax and PowerShell. We're heads down on the v1.1 release, and then I'll have more time to dedicate to fleshing this out. We have been discussing the rich object model needed for a good PowerShell experience.

Reply:
The verb choice is a vital component of the sacred vow. IT Pros would expect to use certain verbs and certain parameter names names. To create a truly  great Powershell module means using approved verbs. :-)